domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/secretbackend"
+	backenderrors "github.com/juju/juju/domain/secretbackend/errors"
+	"github.com/juju/juju/internal/database"
+)
+
+type upsertOperation struct {
+	secretbackend.UpsertSecretBackendParams
+
+	getBackendStmt *sqlair.Statement
+
+	upsertBackendStmt                 *sqlair.Statement
+	upsertRotationStmt                *sqlair.Statement
+	upsertConfigStmt, clearConfigStmt *sqlair.Statement
+}
+
+// build builds the sqlair statements.
+func (o *upsertOperation) build() (err error) {
+	// validate the input parameters here, so we can fail early without accessing the database.
+	if o.ID == "" {
+		return fmt.Errorf("%w: ID is missing", backenderrors.NotValid)
+	}
+	for k, v := range o.Config {
+		if k == "" {
+			return fmt.Errorf(
+				"%w: empty config key for %q", backenderrors.NotValid, o.ID)
+		}
+		if v.(string) == "" {
+			return fmt.Errorf(
+				"%w: empty config value for %q", backenderrors.NotValid, o.ID)
+		}
+	}
+
+	o.getBackendStmt, err = sqlair.Prepare(`
+SELECT 
```

Comment:
I don't think any of these statements need to explicitly list columns. I believe most, if not all, can deal in terms of `SecretBackend.*` for example.